Randomchating

Doncham opened the pull request:
레디스를 통해 사용자 세션을 관리하며 서로 다른 환경에서의 웹소켓 접속을 가정하며 구현을 하기는 했는데 아직 서버 환경에서는 테스트를 안해봤습니다. 
레디스의 리스트를 큐로 활용하여 채팅 대기방을 구현하였고 Spring Retry를 사용하여 15초 동안 1초에 1회 큐를 확인하여 매칭 상대가 있는지 확인하는 로직을 구현하였습니다.
매칭이 되면 매칭 메시지를 양쪽의 웹소켓에 전달하고, 메세지를 서로 보내는 기능을 구현하였습니다.
좀 읽기 힘든 부분이 많아서 ChatRoomService 부분에는 주석을 많이 달아놨습니다. 
테스트 코드는 못 만들었습니다 ㅎ..
Closed #7

Doncham committed:
feat: 엔티티 생성

Doncham committed:
feat: 임시 저장

Doncham committed:
feat: 채팅 입장 메시지 퇴장 메시지 보내기

Doncham committed:
feat: 메시지 전송 로직

Doncham committed:
feat: 매칭 재시도 로직

15초동안 1초에 한 번씩 매칭을 시도한다.
레디스에 저장되는 유저세션 정보에 isMatched를 추가해서 조잡하게 발생되는 오류들을 처리

f-lab-ben commented on RanChat/build.gradle:
```diff
@@ -42,7 +42,13 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-data-redis'
+    implementation 'org.springframework.boot:spring-boot-starter-websocket'
+    implementation 'org.springframework.session:spring-session-data-redis'
+    implementation 'com.fasterxml.jackson.core:jackson-databind'
```
따로 사용하신 이유가 있을까요?
jackson같은 경우 spring core에 기본적으로 import되어있는데요~

f-lab-ben commented on RanChat/build.gradle:
```diff
@@ -42,7 +42,13 @@ dependencies {
     testRuntimeOnly 'com.h2database:h2'
+
+    implementation 'org.springframework.retry:spring-retry'
+    implementation 'org.springframework:spring-aspects'
```
요것도 따로 import 한 이유가 있나요?

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/chatroom/controller/ChatRoomController.java:
```diff
@@ -0,0 +1,23 @@
+public class ChatRoomController {
+    private final ChatRoomService chatRoomService;
+    @PostMapping("ranchat")
+    public ResponseEntity<String> ranchat(@RequestParam String userId) {
```
response의 형태가 너무 raw형태입니다
기본적으로 string 하나만 가지고 있더라도 response dto를 만들어서 return 하는 습관을 들이면 좋아요~

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/chatroom/controller/RanChatResponse.java:
```diff
@@ -0,0 +1,17 @@
+package com.example.ranchat.chatroom.controller;
+
+
+public class RanChatResponse {
```
요건 어째 lombok을 안쓰고 직접구현인가요~?
컨벤션이 통일이 되어야 좋을것 같아요

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/chatroom/entity/userchatroom/UserChatRoom.java:
```diff
@@ -0,0 +1,25 @@
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @MapsId("userKey")
+    @JoinColumn(name = "user_key")
```
조인하는 방법이 여러가지가 있는데
mapsId를 쓴 이유가 있을까요?

Doncham replied:
User 엔티티와 ChatRoom의 다대다 관계를 표현하기 위해서(현재의 랜덤 채팅에서는 1대1이지만 확장성을 위해서 다대다로 설정) 중간 테이블로 UserChatRoom 엔티티를 만들었고, User와 ChatRoom의 기본키를 외래키로 가지고, 둘을 복합 기본키로 만들기 위해서 mapsId를 사용하였습니다. 
IdClass vs EmbeddedId 에서 왜 EmbeddedId을 선택했는지에 대한 내용일까요? 그렇다면 큰 고민 없이 기억에 나는 방법으로 EmbeddedId을 선택하였습니다... 추후에 더 공부해보겠습니다

f-lab-ben replied:
후자는 아니구요~

mapsId를 붙히지않아도 동작할텐데, 사용한 이유에 대해 여쭤보는거였습니다

이미 써봐서 알겠지만 fk를 사용해서 jpa 연관관계를 한번에 다 표현하려고 해보면
여러 어노테이션들을 붙히면서 조회할 수 있게 만들어야하는데

예컨데
Entity A -> 1: N
Entity B -> N : M

이라고했을 때, A가 여러개의 B를 가진다면 어떻게될까요?
조회성능은 또 어떨까요?

여러 문제가 있기에 현업에서는 fk를 사용하지 않고, id만 들고있는 채로 join을 통해 불러오는 방식을 사용합니다
(jpql, querydsl)

지금이야 테이블이 얼마없고 관계 표현이 비교적 간단하지만,
점차 많아지고 복잡해질수록 여기엔 mapsId인지, joinColumn인지.. 등 상당히 복잡하거든요.

꼭 필요한 경우가 아니면 굳이 jpa에 있는 어노테이션들을 다 사용해서 무리한 관계표현을 하는건 성능/가독성 측면에서도 딱히 좋지가 않습니다

아마 상원님도 한달뒤에 상원님코드 보면 이 entity 조회할때 어떤 쿼리가 날아갈지 예측이 안되실거에요 🙂

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/chatroom/entity/ChatRoom.java:
```diff
@@ -0,0 +1,19 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long chatRoomKey;
+
+    private Integer isDeleted;
```
컬럼명을 지정하지 않으면 어떤 경우가 발생할지 추측할 수 있을까요?

isDeleted 같은 경우 테이블의 어떤 컬럼명으로 참조를할까요?

만약 camel case의 테이블을 참조해야한다면 어떻게될까요?

Doncham replied:
컬럼명을 지정하지 않으면 JPA에서 기본 네이밍 전략에 따라 카멜 케이스의 필드를 스네이크 케이스로 변환하여 테이블의 컬럼명으로 매핑할 것 같습니다.
isDelete의 경우 is_deleted가 될 것이고 만약 camel case의 테이블을 참조하는 경우 JPA의 ddl-auto 설정을 none으로하여 스키마 검증을 하지 않은 경우, 런타임에 해당 엔티티를 조회하거나 저장할 때 예외가 발생할 것 같습니다. 
만약 camel case의 테이블을 참조해야하면 hibernate의 경우 PhysicalNamingStrategy를 구현하여 기본 네이밍 전략을 변경하여 카멜케이스 필드를 스네이크 케이스로 변환하지 않고 테이블의 컬럼과 매핑할 수 있을 것 같습니다.

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/chatroom/service/ChatRoomService.java:
```diff
@@ -0,0 +1,131 @@
+        String otherUserId = redisService.getUserFromWaitingRoom(WAITING_QUEUE);
+
+        // 상대가 나를 매칭했는지 확인
+        if ("true".equals(redisService.isMatched(userId))) {
```
isMatched가 어째 string 비교가 되었을까요?
boolean 비교로 최대한 해야 좋습니다.

string 과 boolean의 비교중 어떤게 더 저렴할까요?

Doncham replied:
boolean은 메모리에서 해당 데이터의 비트값이 0인지 1인지만 파악합니다.

String.equals()는 타입이 String인지 체크하고, null 값인지 체크하고, 두 문자열의 길이를 비교한 후 for문을 순회하며 각 문자의 byte값을 비교합니다.

boolean 값으로 바꾸겠습니다

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/chatroom/service/ChatRoomService.java:
```diff
@@ -0,0 +1,131 @@
+                redisService.pushUserToWaitingRoom(WAITING_QUEUE,otherUserId);
+
+            // 매칭이 된 상태니까 재시도 로직을 실행할 필요가 없어
+            return userId + " is matched";
```
너무 날 코딩인데요 이건~

f-lab-ben replied:
매칭 상태와 메세지를 분리한 dto가 있어야겠네요

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/chatroom/service/ChatRoomService.java:
```diff
@@ -0,0 +1,131 @@
+            UserSessionInfo otherUserInfo = redisService.getUserSessionInfo(otherUserId);
+
+            if (otherUserInfo != null) {
+                // 채팅방 ID 생성
```
이 아래 부분은 atomic하게 수행되지 않아도 되나요?
트랜잭션이 없어보이는데요~

Doncham replied:
redis의 트랜잭션을 공부하다가 레디스에 대한 조회 연산이 트랜잭션 내부에 존재하면 null값을 반환한다는 특징을 알았고 그 문제에 대한 해결책으로 Lua script, watch를 통한 낙관적락을 적용해보려고 했으나 Lua script는 유지보수성이 너무 떨어져 보였고 watch 연산을 사용하려고 하니 기존 로직이 너무 복잡해서 연산 사이사이에 다른 클라이언트의 연산이 끼어들 수 있는 상황을 막기 힘들었습니다. 그래서 기존 로직을 주기적으로 @Scheduled를 통해 큐에 사이즈를 확인하여 매칭해주는 방식으로 리팩토링했습니다.

f-lab-ben replied:
좋습니다.

scheduled 를 사용하면 주기적으로 동작하는 polling인데,
다른 방법으론 event 방식인 경우도 있을거에요. redis pub/sub이나 kafka같은 방식과 비교해보고

어떤 방식을 써야하는지 머릿속에 기억해주면 좋을거같아요~

Doncham replied:
이벤트 방식으로 구현하는게 트래픽이 적은 경우에는 불필요한 연산을 줄일 수 있겠네요!!!

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/chatroom/service/ChatRoomService.java:
```diff
@@ -0,0 +1,131 @@
+        try {
+            String message = objectMapper.writeValueAsString(messageDTO);
+            // 레디스에 보내면 다른 서버까지 메세지가 전송됨.
+            redisPublisher.publish("chatRoom:" + chatRoomId, message);
```
메세지를 serialize해서 보내는이유가 있나요?
object 째로 보내면 redis serializer가 serialize 해주는데 이렇게 한 이유가 궁금합니다

Doncham replied:
redisTemplate을 커스텀하지 않고, Spring Boot에 의해서 활성화되는 redisTemplate 빈을 사용하다 보니까 Serializer에 대한 개념이 부족해서 직접 직렬화 해줘야 하는 줄 알았습니다.

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/redis/RedisPublisher.java:
```diff
@@ -0,0 +1,33 @@
+
+    private final RedisTemplate<String, String> redisTemplate;
+
+    public void publish(String channel, String message) {
```
message가 string일 경우 
이 메서드의 경우 거의 Object를 받는것과 동일한 형태일 것 같아요

어떤 단점이 있을까요? 테스트는 어떻게 해야할까요?

f-lab-ben replied:
redis는 spring에서 사용하는 방법은 별로 중요치가 x

redis-cli에서 어떤 프로토콜로 어떻게 동작하는지가 제일 중요

leftPush rightPush
-> list

LPUSH, RPUSH
LPOP, RPOP

-> 어떤 파라미터를 넘겨줘야하고, 리턴값은 뭔지
그리고 시간복잡도가 얼만큼인지

LMEMBER
-> O(1)

전체 있는 key를 조회하는 keys()
-> O(N)

--> 이게 스프링에서는 어떤 메서드로 추상화되었는가?

f-lab-ben replied:
python, go, javascript

leftPush rightPush

lpop, lpush

f-lab-ben replied:
use-case가 어떻게 되는가?

f-lab-ben replied:
redis는 그냥 java에서 지원하는거 다 있음
-> 이거를 그냥 분산해서 저장하고, 여러 서버가 접근해도 동시성문제없는 자료구조

f-lab-ben replied:
list
add()

Queue <<
put, get << Map

zset << TreeMap

f-lab-ben replied:
서버 자체 : redis cluster

클라이언트 : lettuce 

서버의 데이터 저장 : memory, file(async)

Doncham replied:
테스트를 작성할 때 메세지 작성 시 객체로 작성하지 못하고 
String jsonMessage = "{\"chatRoomId\":\"123\",\"content\":\"Hello\",\"webSocketSessionId\":\"sessionId\",\"sender\":\"user1\"}";
이런 식으로 메세지를 작성해야해서 테스트가 힘들 것 같습니다.
동료 개발자들도 어떤 데이터를 넣어야할지 모르는 상황이 생길 수 있어서 DTO를 새로 만드는게 좋은거 같습니다.

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/redis/RedisPublisher.java:
```diff
@@ -0,0 +1,33 @@
+        // 메시지를 파싱하여 채널 ID(chatRoomId)를 추출하는 로직을 구현합니다.
+        // JSON 파싱을 위해 ObjectMapper 등을 사용할 수 있습니다.
+
+        return "chatRoom:"; //+ parsedChatRoomId;
```
이런 하드코딩류는 util 클래스 등으로 빼면 좋겠죠?

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/redis/RedisSubscriber.java:
```diff
@@ -0,0 +1,65 @@
+        log.info("body가 어떤 형태이지?: " + body);
+
+        try {
+            MessageDTO messageDTO = objectMapper.readValue(body, MessageDTO.class);
```
redis serializer 등을 사용하는 방법을 찾아보는게 좋을것 같아요.
이렇게되면 type 안정성이 없는채로 동작하고, 이후에 어떤 클래스가 매핑되는지 알수가없을거에요

Doncham replied:
네 Serializer를 통해서 MessageDTO로 변경하는 방식으로 변경하였습니다.

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/redis/RedisSubscriber.java:
```diff
@@ -0,0 +1,65 @@
+                for (String userId : userIds) {
+                    WebSocketSession session = sessionManager.getSession(userId);
+                    if (session != null && session.isOpen()) {
+                        synchronized (session) {
```
동기화하는 방법 중 가장 성능이 안좋은 방법인데요, 다른 방법은 없을까요?
유저 id가 많아지면 어떻게될까요?

Doncham replied:
ReentrantLock으로 변경했습니다. 유저Id가 많아지면 락 획득을 위한 경합이 많이 발생하는데 JVM이 락 획득을 관리하는데 많은 오버헤드가 발생할 것 같습니다.

f-lab-ben replied:
좋습니다~

실제로 spring 진영에서도 synchronized로 되어있는 레거시 코드들이 많은데
점진적으로 ReentrantLock으로 변경중입니다.

성능차이가 유의미하게 크게나니, 꼭 현업에선 사용하지 않도록 해주세요~

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/redis/Service/RedisService.java:
```diff
@@ -1,27 +1,92 @@
+        String key = "userSession:" + info.getUserId();
+        redisTemplate.opsForHash().put(key,"userId", info.getUserId());
+        redisTemplate.opsForHash().put(key,"webSocketSessionId", info.getWebSocketSessionId());
+        redisTemplate.opsForHash().put(key,"isMatched", "false");
```
false로 넣던지

혹은 dto 형태로 말아넣던지 하는 대책이 필요해보여요

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/redis/Service/RedisService.java:
```diff
@@ -1,27 +1,92 @@
+
+        if (webSocketSessionId != null) {
+            UserSessionInfo info = new UserSessionInfo();
+            info.setUserId(userId);
```
setter 지양해주세요~
왠만하면 생성자에 넣읍시다 

혹은 메서드로 새로 작성을 하던지요~

Doncham replied:
빌더패턴으로 변경했습니다

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/redis/Service/RedisService.java:
```diff
@@ -1,27 +1,92 @@
+    }
+
+    public List<String> getUserIds(String chatRoomId) {
+        String userId1 = (String)redisTemplate.opsForHash().get(chatRoomId, "userId1");
```
여긴 하드코딩된건가요?

Doncham replied:
채팅방에 있는 유저들의 저장하기 위한 부분입니다. 
Key: chatRoomId:123
Fields:
   - "userId1" : 1
   - "userId2" : 2
이런 식으로 저장했습니다

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/chatroom/service/ChatRoomService.java:
```diff
@@ -0,0 +1,131 @@
+                makeMatchState(userId, otherUserId);
+
+                // 두 사용자에게 채팅방 입장 메시지 전송
+                sendMatchNotification(currentUserInfo, chatRoomId, currentUserInfo.getWebSocketSessionId());
```
메세지가 점점 많아질텐데, 계속 네이밍을 바꾸면서 메서드를 수정하는건 메서드가 너무많아질것 같은데요
enum으로 메세지 type을 정해서 호출하면 어떨까요?

Doncham replied:
MessageDTO에 Enum으로 type필드를 만들었습니다!

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/chatroom/service/ChatRoomService.java:
```diff
@@ -0,0 +1,131 @@
+        // 현재 사용자 정보 가져오기, 웹소켓 세션Id를 통해 메세지 전송
+        UserSessionInfo currentUserInfo = redisService.getUserSessionInfo(userId);
+
+        // 매칭 시도, 큐에서 하나 뽑기
```
매칭하는 api인데 이거 최악의 경우 계속 api를 기다리다가 딜레이 되는 형태겠네요.

비동기로 클라이언트에게 알려주는 방식을 택하지않은 이유가 있나요?

A -> 매칭 요청 -> return 후 대기
B -> 매칭 요청 -> return 후 대기

서버에선 내부적으로 매칭되는 유저들 주기적으로 체크하거나, 
매칭요청 발생 시 체크해서 매칭되는 유저들에게 websocket으로 노티

이런 시퀀스가 좀더 낫지 않을까요?

Doncham replied:
매칭 api 실행 시 바로 응답을 반환하도록 설정했습니다. @Retry + @Async를 통해 매칭 로직을 백그라운드에서 수행하게 했고 매칭 성공 시 웹소켓을 통해 연결 정보를 전송하도록 했습니다.

f-lab-ben replied:
async를 사용해서 돌리면 비동기로 돌게 되지만,
예컨데 서버 재시작이나 배포하는 경우에는 어떻게 될까요?

그럴땐 가지고있는 데이터를 다 잃고 매칭중이지만, 실제로 매칭은 안될거에요.
이럴때도 특정 redis queue에 넣고 주기적 폴링을 하던지 pub sub 하는 방향으로 가도 좋아요

Doncham replied:
확실히 서버 재시작할 때 기존의 돌던 로직이 어떻게 될지는 전혀 생각해보지 못한 관점입니다!!

f-lab-ben reviewed (commented):
PR 리뷰 확인해주세요~~

Doncham committed:
feat: 리팩토링

1.RedisService 역할 분리
2.ChatRoomController 응답 DTO 생성
3.Value 타입별 redisTemplaet 빈 생성

Doncham committed:
feat: 재시도 로직 -> 재시도 + 비동기 로직

@Async + @Retry 로 매칭 로직 리팩토링

Doncham committed:
feat: messageDTO를 전송

(1)messageDTO를 역직렬화하기 위해 messageListenrAdapter에 Serializer 추가
(2)RedisSubscriber에서 lock을 reentrantLock으로 구현. webSocketSession별로 reentrantLock을 하나씩 만들어서 동기화 구현

Doncham committed:
feat: 매칭 로직 리팩토링

(1) Retry를 통한 매칭 로직 -> Scheduler를 사용한 주기적인 매칭 처리
(2) Redisson을 통해 여러 서버에서 동시에 매칭 처리를 하지 못하도록 동기화
(3) 메시지 type 지정

Doncham committed:
feat: test 추가

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/chatroom/entity/userchatroom/UserChatRoom.java:
```diff
@@ -0,0 +1,25 @@
+
+@Entity
+@NoArgsConstructor
+public class UserChatRoom extends BaseEntity{
```
전체적으로 indentation이 잘 안맞는것 같아요.

https://bestinu.tistory.com/64

참고해서 formatter 적용해서 사용해보세요~

f-lab-ben commented:
머지해도 좋겠습니다~ 👍

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/chatroom/service/ChatRoomService.java:
```diff
@@ -0,0 +1,74 @@
+            }
+            Long size = redisService.checkQueueSize(WAITING_QUEUE);
+
+            String userAId,userBId;
```
이런식 네이밍안좋아요~

항상 코드짤때 a,b,c ... 이런식이거나 1,2,3 등의 시퀀셜한 무언가는 피해보세요

그리고 변수선언 하고 할당을 나중에 하는 식이면
userA,BId는 언제든 또 값이 바뀔 수 있겠죠?

최대한 setter 사용하지 않는 방향처럼
변수는 선언하는 시점에 초기화를 하고, 이후엔 변경이안되도록 사용하는게 제일 좋습니다

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/chatroom/service/ChatRoomService.java:
```diff
@@ -0,0 +1,74 @@
+                log.info("Not enough for matching");
+            }
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
```
로깅안하면 원인찾을수없을거에요
여긴 스프링 request scope이아니어서 advice도 안먹힐거고
로깅도 아예없으면 디버깅이 안되지않을까요?

Doncham replied:
앗 당연히 ExceptionHandler에서 잡을 수 있는 줄 알았습니다...

f-lab-ben commented on RanChat/src/main/java/com/example/ranchat/config/AsyncConfig.java:
```diff
@@ -0,0 +1,25 @@
+@Configuration
+public class AsyncConfig {
+    @Bean
+    public Executor taskExecutor() {
```
좋아요 👍

f-lab-ben replied:
이렇게되면 공통 taskThreadPool인건데,
bean이름을 명시해서 사용하면 좋겠죠?

async 어노테이션 명시할때 thread pool을 지정할 수 있습니다

Doncham committed:
feat: 리팩토링

네이버 컨벤션 적용, 네이밍 수정
